keras/layers/recurrent_convolutional.py

```diff
@@ -0,0 +1,517 @@
+                 input_dim=None, input_length=None, **kwargs):
+        self.return_sequences = return_sequences
+        self.initial_weights = weights
+        self.go_backwards = go_backwards
+        self.stateful = stateful
+
+        self.nb_row = nb_row
+        self.nb_col = nb_col
+        self.nb_filter = nb_filter
+        self.dim_ordering = dim_ordering
+        self.input_spec = [InputSpec(ndim=5)]
+
+        self.input_dim = input_dim
+        self.input_length = input_length
+        #if self.input_dim:
+        #    kwargs['input_shape'] = (self.input_length, self.input_dim)
+
+        super(RecurrentConv2D, self).__init__(**kwargs)
+
+    def compute_mask(self, input,mask):
+        if self.return_sequences:
+            return mask
+        else:
+            return None
+
+    def get_output_shape_for(self,input_shape):
+
+        if self.dim_ordering == 'th':
+            rows = input_shape[2+1]
+            cols = input_shape[3+1]
+        elif self.dim_ordering == 'tf':
+            rows = input_shape[1+1]
+            cols = input_shape[2+1]
+        else:
+            raise Exception('Invalid dim_ordering: ' + self.dim_ordering)
+
+        rows = conv_output_length(rows, self.nb_row,
+                                  self.border_mode, self.subsample[0])
+        cols = conv_output_length(cols, self.nb_col,
+                                  self.border_mode, self.subsample[1])
+
+        if self.return_sequences:
+            if self.dim_ordering == 'th':
+                return (input_shape[0], input_shape[1],
+                        self.nb_filter, rows, cols)
+            elif self.dim_ordering == 'tf':
+                return (input_shape[0], input_shape[1],
+                        rows, cols, self.nb_filter)
+            else:
+                raise Exception('Invalid dim_ordering: ' + self.dim_ordering)
+        else:
+            if self.dim_ordering == 'th':
+                return (input_shape[0], self.nb_filter, rows, cols)
+            elif self.dim_ordering == 'tf':
+                return (input_shape[0], rows, cols, self.nb_filter)
+            else:
+                raise Exception('Invalid dim_ordering: ' + self.dim_ordering)
+
+    def step(self, x, states):
+        raise NotImplementedError
+
+    def get_constants(self, X, train=False):
+        return None
+
+    def get_initial_states(self, X):
+        # (samples, timesteps, row, col, filter)
+        initial_state = K.zeros_like(X)
+        # (samples,row, col, filter)
+        initial_state = K.sum(initial_state, axis=1)
+        # initial_state = initial_state[::,]
+        initial_state = self.conv_step(initial_state, K.zeros(self.W_shape),
+                                       border_mode=self.border_mode)
+
+        initial_states = [initial_state for _ in range(2)]
+        return initial_states
+        
+    def preprocess_input(self, x):
+        return x
+    def call(self, x,mask=None):
+
+       
```

Comment:
too many white spaces. You only need one white space between functions inside a class